File: estimator/kalman_filter.c
// Kalman Filter implementation in C
#include "kalman_filter.h"

void kalman_init(KalmanFilter1D* kf, float A, float B, float H, float Q, float R) {
    kf->A = A;
    kf->B = B;
    kf->H = H;
    kf->Q = Q;
    kf->R = R;
    kf->P = 1.0f;  // Initial error covariance
    kf->x = 0.0f;  // Initial state
}

void kalman_predict(KalmanFilter1D* kf, float u) {
    kf->x = kf->A * kf->x + kf->B * u;
    kf->P = kf->A * kf->P * kf->A + kf->Q;
}

float kalman_update(KalmanFilter1D* kf, float z) {
    float y = z - kf->H * kf->x;
    float S = kf->H * kf->P * kf->H + kf->R;
    float K = kf->P * kf->H / S;

    kf->x = kf->x + K * y;
    kf->P = (1 - K * kf->H) * kf->P;

    return kf->x;
}

File: estimator/kalman_filter.py

import numpy as np

class KalmanFilter1D:
    def __init__(self, A, B, H, Q, R, P, x0):
        self.A = A  # State transition
        self.B = B  # Control input
        self.H = H  # Observation
        self.Q = Q  # Process noise
        self.R = R  # Measurement noise
        self.P = P  # Initial estimation error
        self.x = x0  # Initial state

    def predict(self, u):
        self.x = self.A @ self.x + self.B * u
        self.P = self.A @ self.P @ self.A.T + self.Q
        return self.x

    def update(self, z):
        y = z - self.H @ self.x
        S = self.H @ self.P @ self.H.T + self.R
        K = self.P @ self.H.T @ np.linalg.inv(S)
        self.x += K @ y
        self.P = (np.eye(self.P.shape[0]) - K @ self.H) @ self.P
        return self.x
